File: test/com/barrybecker4/optimization/optimizee/optimizees/TravelingSalesmanProblem.java
/** Copyright by Barry G. Becker, 2000-2011. Licensed under MIT License: http://www.opensource.org/licenses/MIT  */
package com.barrybecker4.optimization.optimizee.optimizees;

import com.barrybecker4.optimization.Optimizer;
import com.barrybecker4.optimization.parameter.ParameterArray;
import com.barrybecker4.optimization.strategy.OptimizationStrategyType;

import static com.barrybecker4.optimization.OptimizerTestCase.LOG_FILE_HOME;

/**
 * This is a simple search space to test the optimization package.
 *
 * @author Barry Becker
 */
public class TravelingSalesmanProblem extends OptimizeeProblem {

    private TravelingSalesmanVariation variation_ = TravelingSalesmanVariation.SIMPLE;


    /** constructor */
    public TravelingSalesmanProblem(TravelingSalesmanVariation variation) {
        variation_ = variation;
    }

    @Override
    public String getName() {
        return "Traveling Salesman Problem";
    }

    /**
     * we evaluate directly not by comparing with a different trial.
     */
    @Override
    public boolean evaluateByComparison() {
        return false;
    }

    /**
     * Use the cost matrix for the TSP variation to determine this.
     * @return fitness value
     */
    @Override
    public double evaluateFitness(ParameterArray a) {
        return variation_.evaluateFitness(a);
    }

    @Override
    public ParameterArray getExactSolution() {
        return variation_.getExactSolution();
    }

    @Override
    public ParameterArray getInitialGuess() {
        return variation_.getInitialGuess();
    }

    @Override
    public double getFitnessRange() {
        return variation_.getFitnessRange();
    }

    /**
     * This finds the solution for the above optimization problem.
     */
    public static void main(String[] args) {
        TravelingSalesmanVariation v = TravelingSalesmanVariation.SIMPLE;
        OptimizeeProblem problem = new TravelingSalesmanProblem(v);
        Optimizer optimizer =
                new Optimizer(problem, LOG_FILE_HOME + "tsp_optimization.txt");

        ParameterArray initialGuess = problem.getInitialGuess();

        ParameterArray solution =
                optimizer.doOptimization(OptimizationStrategyType.SIMULATED_ANNEALING, initialGuess, v.getFitnessRange());

        showSolution(problem, solution);
    }
}

File: scala-test/com/barrybecker4/optimization/optimizee/optimizees/problems/TravelingSalesmanProblem.scala
// Copyright by Barry G. Becker, 2000-2018. Licensed under MIT License: http://www.opensource.org/licenses/MIT
package com.barrybecker4.optimization.optimizee.optimizees.problems

import com.barrybecker4.optimization.Optimizer
import com.barrybecker4.optimization.optimizee.OptimizerTestSuite.LOG_FILE_HOME
import com.barrybecker4.optimization.optimizee.optimizees.OptimizeeProblem
import com.barrybecker4.optimization.parameter.{ParameterArray, ParameterArrayWithFitness}
import com.barrybecker4.optimization.strategy.SIMULATED_ANNEALING
import OptimizeeProblem.showSolution

import scala.util.Random


object TravelingSalesmanProblem {

  /** Finds the solution for the above optimization problem. */
  def main(args: Array[String]): Unit = {
    val v = TSP_SIMPLE // TSP_STANDARD // TSP_US_CAPITALS // TSP_SIMPLE
    val problem = new TravelingSalesmanProblem(v)
    val optimizer = new Optimizer(problem, Some(LOG_FILE_HOME + "tsp_optimization.txt"))
    val initialGuess = problem.getInitialGuess

    val solution = optimizer.doOptimization(SIMULATED_ANNEALING,
      initialGuess, v.getFitnessRange, new Random(1))

    showSolution(problem, solution)
  }
}

/**
  * This is a simple search space to test the optimization package.
  * @author Barry Becker
  */
class TravelingSalesmanProblem(val variation: TravelingSalesmanVariation = TSP_SIMPLE)
    extends OptimizeeProblem {

  override def getName: String = "TSP: " + variation.getName

  /** Evaluate directly not by comparing with a different trial */
  override def evaluateByComparison = false

  /** Use the cost matrix for the TSP variation to determine this.
    * @return fitness value
    */
  override def evaluateFitness(a: ParameterArray): Double = variation.evaluateFitness(a)

  override def getExactSolution: ParameterArrayWithFitness = variation.getExactSolution

  override def getInitialGuess: ParameterArray = variation.getInitialGuess

  override def getFitnessRange: Double = variation.getFitnessRange
}
